pkg/controllers/provisioning/scheduling/scheduler.go

```diff
@@ -16,40 +16,43 @@ package scheduling
 
 import (
 	"context"
+	"fmt"
 	"sort"
 
-	"github.com/samber/lo"
-
-	"github.com/aws/karpenter/pkg/events"
-	"github.com/aws/karpenter/pkg/scheduling"
-
-	"github.com/aws/karpenter/pkg/controllers/state"
-
 	"go.uber.org/multierr"
 	v1 "k8s.io/api/core/v1"
 	"knative.dev/pkg/logging"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 
 	"github.com/aws/karpenter/pkg/apis/provisioning/v1alpha5"
 	"github.com/aws/karpenter/pkg/cloudprovider"
+	"github.com/aws/karpenter/pkg/controllers/state"
+	"github.com/aws/karpenter/pkg/events"
+	"github.com/aws/karpenter/pkg/scheduling"
+	"github.com/aws/karpenter/pkg/utils/resources"
 )
 
 func NewScheduler(nodeTemplates []*scheduling.NodeTemplate, cluster *state.Cluster, topology *Topology,
 	instanceTypes []cloudprovider.InstanceType, daemonOverhead map[*scheduling.NodeTemplate]v1.ResourceList, recorder events.Recorder) *Scheduler {
 	sort.Slice(instanceTypes, func(i, j int) bool { return instanceTypes[i].Price() < instanceTypes[j].Price() })
 	s := &Scheduler{
-		nodeTemplates:  nodeTemplates,
-		topology:       topology,
-		cluster:        cluster,
-		instanceTypes:  instanceTypes,
-		daemonOverhead: daemonOverhead,
-		recorder:       recorder,
-		preferences:    &Preferences{},
+		nodeTemplates:      nodeTemplates,
+		topology:           topology,
+		cluster:            cluster,
+		instanceTypes:      instanceTypes,
+		daemonOverhead:     daemonOverhead,
+		recorder:           recorder,
+		preferences:        &Preferences{},
+		remainingResources: map[string]v1.ResourceList{},
```

Comment:
Thoughts on inverting this to `resources`?

Reply:
If it's just the resources used, then you need to also store the provisioner limits so you can calculate.  It would work, but it's an extra variable and this way you just need to filter for instance types smaller than the remaining resources which is pretty simple as opposed to calculating that amount each time and then filtering.